Tag entity exclusion

alek-kam-robotec-ai opened the pull request:
This pr adds a new way to exclude entities from raycasting using the Tag component.
 - excluded tag names were added as a field to the scene configuration component.
 - added a way to dynamically handle entity exclusion from raycasting based on their tags.  
 
 To exclude an entity from raycasting tag it using the tag component and include the tag name in the scene configuration componet.

msz-rai commented on Code/Source/RGLSystemComponent.cpp:
```diff
@@ -121,14 +135,37 @@ namespace RGL
+            return false;
+        }
+
+        for (const auto tag : excludedTags)
+        {
+            if (entityTags.contains(tag))
+            {
+                return true;
+            }
+        }
+
+        return false;
```
```suggestion
        return std::any_of(
            excludedTags.cbegin(),
            excludedTags.cend(),
            [&entityTags](const auto& tag)
            {
                return entityTags.contains(tag);
            });
```

msz-rai commented on Code/Source/RGLSystemComponent.cpp:
```diff
@@ -189,6 +232,26 @@ namespace RGL
         m_modelLibrary.Clear();
     }
 
+    void RGLSystemComponent::OnTick(float deltaTime, AZ::ScriptTimePoint time)
+    {
+        for (auto entityId : m_managersToBeRemoved)
+        {
+            m_entityManagers.erase(entityId);
+            m_unprocessedEntities.insert(entityId);
+        }
+        m_managersToBeRemoved.clear();
+    }
```
Why couldn't we erase entity manager in `ReviseEntityPresence(AZ::EntityId entityId)`?

alek-kam-robotec-ai replied:
This is so that the caller can signal revision inside a bus notification and can then safely disconnect from said bus after the call has ended (on destruction).

msz-rai replied:
At the moment, the only caller is `EntityTagListener` which is not destroyed in `ReviseEntityPresence()` so your case doesn't occur. Is it solution for the future usages?

alek-kam-robotec-ai replied:
Exactly. I believe that this could lead to bugs that would be hard to track down as in previous iterations of the solution the entity manager handlet the tag notification bus and the `ReviseEntityPresence` method was called from within the callbacks which lead to segfaults caused by destruction of the entity manager before the callback was finished. In my opinion a small amount of additional complexity in this case is worth it.

msz-rai commented on Code/Source/Entity/EntityTagListener.h:
Instead of creating the new class and having separate container for it in `RGLSystemComponent` I'd extend `EntityManager` with `LmbrCentral::TagComponentNotificationsBus::Handler`. I see no advantages in having this separation.

alek-kam-robotec-ai replied:
This is to allow for a scenario where an entity is excluded from the raycasting due it being tagged with an excluded tag after which the tag is removed and the entity must be processed. With the bus handled by the entity manager we would not be able to achieve this
.

msz-rai replied:
I wonder if dynamic handling of tag creation/destruction is necessary (if there are any real use cases). If not, this solution could be significantly simplified to checking `ShouldEntityBeExcluded` in `OnEntityContextCreateEntity` and discarding entities that should be excluded (skipping processing them and putting them to `m_unprocessedEntities`). Did you get the detailed requirements for this feature?

alek-kam-robotec-ai replied:
The Tag bus is able to handle tag addition and removal at runtime. I believe we should cover these cases as some proved useful before (as seen in [this pr](https://github.com/RobotecAI/o3de-rgl-gem/pull/53)).

msz-rai reviewed (changes requested):
Good job :+1: 

When testing I got the following bug:
1. I added cube with tag that is excluded in scene configuration.
2. I ran the simulation and the cube was ignored by RGL :+1: 
3. I stopped the simulation and removed tag from the cube (Tag component is still present)
4. I ran the simulation again and the cube was ignored by RGL even though the tag was removed.

Please investigate my test case. Also, I left some comments to address.

alek-kam-robotec-ai commented:
> Good job 👍
> 
> When testing I got the following bug:
> 
> 1. I added cube with tag that is excluded in scene configuration.
> 2. I ran the simulation and the cube was ignored by RGL 👍
> 3. I stopped the simulation and removed tag from the cube (Tag component is still present)
> 4. I ran the simulation again and the cube was ignored by RGL even though the tag was removed.
> 
> Please investigate my test case. Also, I left some comments to address.

@msz-rai
This must be a bug in the engine. I read the entity's tag list on demand. I've checked this and the tag component lists the excluded tag after edition in the editor.

msz-rai commented on Code/Source/RGLSystemComponent.cpp:
```diff
@@ -121,14 +135,35 @@ namespace RGL
         }
 
-        if (m_activeLidarCount < 1U)
+        m_entityTagListeners.emplace_back(entity.GetId());
```
`EntityTagListeners` are never erased. Their destruction should be in `OnEntityContextDestroyEntity()` method.

msz-rai reviewed (approved):
LGTM :+1:
